{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3ba7bff7",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import time\n",
    "\n",
    "from typing import Dict, List\n",
    "from Bio import Entrez\n",
    "from pathlib import Path\n",
    "import xml.etree.ElementTree as ET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "182bd10b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# setup\n",
    "def setup_result_directory(result_dir:str)->str:\n",
    "    '''\n",
    "    Creates a directory by using a specified string as argument.\n",
    "\n",
    "    :param\n",
    "        result_dir: directory path as string\n",
    "    :return:\n",
    "        result_dir_path: Path based on the specified result_dir string argument.\n",
    "    '''\n",
    "    try:\n",
    "        \n",
    "        result_dir_path = result_dir\n",
    "        if os.path.isdir(result_dir_path) == False:\n",
    "            print(\"[*] Trying to create specified directory: {}.\".format(result_dir))\n",
    "            try:\n",
    "                os.mkdir(result_dir)\n",
    "                print(\"\\t[+] DONE\")\n",
    "            except Exception as e:\n",
    "                raise Exception(\"[-] Not able to create directory with exception: {}\".format(e))\n",
    "        else:\n",
    "            print(\"[*] Directory already exist.\")\n",
    "        return result_dir_path\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR creating result directory with exception: {}\".format(e))\n",
    "\n",
    "def setup_entrez(email:str) -> None:\n",
    "    '''\n",
    "    Setup for the biopython Entrez.email field.\n",
    "\n",
    "    :param\n",
    "        email: String variable describing the user email.\n",
    "    :return:\n",
    "    '''\n",
    "    Entrez.email = email"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6e5422d9",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# pipeline step 1 - query NCBI based on a search string\n",
    "def fetch_bioprojects(query:str, max_returns=100)->List[int]:\n",
    "    '''\n",
    "    Function to perform an Entrez search query on the bioproject database.\n",
    "\n",
    "    :param\n",
    "        query: search string\n",
    "    :param\n",
    "        max_returns: maximal number of bioprojects to return\n",
    "    :return:\n",
    "        bioproject_search_results[\"IdList\"]: List[int] of bioproject identifier that can be used for fetching detailed information\n",
    "    '''\n",
    "    try:\n",
    "        bioproject_search_handle = Entrez.esearch(db=\"bioproject\", term=query, retmax=max_returns)\n",
    "        bioproject_search_results = Entrez.read(bioproject_search_handle)\n",
    "        bioproject_search_handle.close()\n",
    "        print(\"[*] Info: Found {} BioProject's associated to Exaiptasia microbiome projects.\".format(bioproject_search_results[\"Count\"]))\n",
    "        return bioproject_search_results[\"IdList\"]\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during fetching bioproject data with exception: {}\".format(e))\n",
    "\n",
    "# pipeline step 2 - get detailed information based on Entrez bioproject identifier\n",
    "def fetch_detailed_bioproject_infos(bioproject_identifier:List[int])->List[Dict]:\n",
    "    '''\n",
    "    Function for extracting detailed information of bioproject entries.\n",
    "\n",
    "    :param\n",
    "        bioproject_identifier: List[int]: list of bioproject identifier\n",
    "    :return:\n",
    "        bioproject_docsum: List[Dict]: detailed information of the requested bioprojects.\n",
    "    '''\n",
    "    try:\n",
    "        to_fetch_ids = \",\".join(bioproject_identifier)\n",
    "        bioproject_docsum_handle = Entrez.efetch(db=\"bioproject\", id=to_fetch_ids, rettype=\"docsum\", retmode=\"xml\")\n",
    "        bioproject_docsum = Entrez.read(bioproject_docsum_handle)[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "        bioproject_docsum_handle.close()\n",
    "        print(\"[*] Info: Found detailed information for: {} associated BioProject's\".format(len(bioproject_docsum)))\n",
    "        return bioproject_docsum\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during fetching bioproject detailed informarion: {}\".format(e))\n",
    "\n",
    "# pipeline step 3 - writing results into CSV file\n",
    "def write_bioproject_result_file(bioprojects_resultfile:str,project_document_summary:List[Dict])->None:\n",
    "    '''\n",
    "    Function for extracting document summary information fields. The information are written into a CSV file.\n",
    "\n",
    "    :param\n",
    "        bioprojects_resultfile: str: path to result file\n",
    "    :param\n",
    "        project_document_summary: List[Dict]: Entrez document summary\n",
    "    :return\n",
    "        None but writes a CSV file\n",
    "    '''\n",
    "    try:\n",
    "        with open(bioprojects_resultfile,\"w\") as biofile:\n",
    "            header = \"ProjectId\\tProjectAcc\\tProjectDate\\tProjectTitle\\tProjectDescription\\tOrganismName\\tOrganismStrain\\n\"\n",
    "            biofile.write(header)\n",
    "            for project in project_document_summary:\n",
    "                project_id = project[\"Project_Id\"]\n",
    "                project_acc = project[\"Project_Acc\"]\n",
    "                project_date = project[\"Registration_Date\"]\n",
    "                project_title = project[\"Project_Title\"]\n",
    "                project_description = project[\"Project_Description\"]\n",
    "                project_organism = project[\"Organism_Name\"]\n",
    "                project_strain = project[\"Organism_Strain\"]\n",
    "                entry = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(project_id,\n",
    "                                                        project_acc,\n",
    "                                                        project_date,\n",
    "                                                        project_title,\n",
    "                                                        project_description,\n",
    "                                                        project_organism,\n",
    "                                                        project_strain)\n",
    "                biofile.write(entry)\n",
    "        print(\"[*] INFO: done writing result file with {} entries.\".format(len(project_document_summary)))\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during BioProject result file writing with exception: {}\".format(e))\n",
    "\n",
    "# pipeline step 4 - fetch biosample information for SRR retrieval\n",
    "def get_biosample_from_bioprojects(bioproject_dataframe:pd.DataFrame)->Dict:\n",
    "    '''\n",
    "    This function reads a bioproject pandas dataframe, extrcats the bioproject identifier and links those to the corresponding\n",
    "    biosamples using the elink function of Entrez. The fetched biosample identifier are stored in a List, which is used as a value\n",
    "    for a dictionary, containing the corresponding bioproject identifiers as keys.\n",
    "\n",
    "    :param\n",
    "        bioproject_dataframe: pd.DataFrame: pandas dataframe of bioprojects, must contain a ProjectId column containing valid bioproject identifier\n",
    "    :return\n",
    "        bioprojects_to_biosamples: Dict: dictionary with bioproject identifier as keys and biosample identifiers as List values\n",
    "    '''\n",
    "    try:\n",
    "        bioprojects_to_biosamples = {}\n",
    "        print(\"[*] Linking bioproject identifier to biosample database.\")\n",
    "        if \"ProjectId\" not in list(bioproject_dataframe.columns):\n",
    "            raise Exception(\"[-] ERROR there is no column called ProjectId in the bioproject dataframe.\")\n",
    "        bioproject_identifier = bioproject_dataframe.ProjectId.apply(lambda identifier: str(identifier)).to_list()\n",
    "        attempts = 3\n",
    "        for project_id in bioproject_identifier:\n",
    "            # try querying NCBI three times\n",
    "            for attempt in range(1, attempts + 1):\n",
    "                try:\n",
    "                    biosample_links = Entrez.elink(dbfrom=\"bioproject\",db=\"biosample\", id=project_id)\n",
    "                    biosample_link_results = Entrez.read(biosample_links)\n",
    "                    biosample_links.close()\n",
    "                    break # exit attempt loop if function succeeded\n",
    "                except Exception as e:\n",
    "                    print(\"\\t[-] ERROR during elink procedure with exception: {}\".format(e))\n",
    "                    if attempt == attempts:\n",
    "                        raise Exception(e)\n",
    "                    else:\n",
    "                        print(\"\\t[*] WARNING: trying elink procedure again after 1 second of sleep ...\")\n",
    "                        time.sleep(1)\n",
    "\n",
    "            biosample_ids = []\n",
    "            for link_set in biosample_link_results:\n",
    "                for link in link_set[\"LinkSetDb\"][0][\"Link\"]:\n",
    "                    link_id = str(link[\"Id\"])\n",
    "                    if link_id not in biosample_ids:\n",
    "                        biosample_ids.append(link_id)\n",
    "            bioprojects_to_biosamples[project_id] = biosample_ids\n",
    "            print(\"\\t[*] Found {} entries for project: {}.\".format(len(biosample_ids), project_id))\n",
    "        print(\"[+] DONE fetching information for biosamples\")\n",
    "        return bioprojects_to_biosamples\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR linking bioprojects to biosamples with exception: {}\".format(e))\n",
    "\n",
    "# pipeline step 5 - fetch SRR data from bioprojects\n",
    "def get_sra_from_bioprojects(bioproject_dataframe:pd.DataFrame)->Dict:\n",
    "    '''\n",
    "    This function reads a bioproject pandas dataframe, extrcats the bioproject identifier and links those to the corresponding\n",
    "    sra entries using the elink function of Entrez. The fetched sra identifier are stored in a List, which is used as a value\n",
    "    for a dictionary, containing the corresponding bioproject identifiers as keys.\n",
    "\n",
    "    :param\n",
    "        bioproject_dataframe: pd.DataFrame: pandas dataframe of bioprojects, must contain a ProjectId column containing valid bioproject identifier\n",
    "    :return\n",
    "        bioprojects_to_sra: Dict: dictionary with bioproject identifier as keys and sra identifiers as List values   \n",
    "    '''\n",
    "    try:\n",
    "        bioprojects_to_sra = {}\n",
    "        print(\"[*] Linking bioproject identifier to srr database\")\n",
    "        if \"ProjectId\" not in list(bioproject_dataframe.columns):\n",
    "            raise Exception(\"[-] ERROR there is no column called ProjectId in the bioproject dataframe.\")\n",
    "        bioproject_identifier = bioproject_dataframe.ProjectId.apply(lambda identifier: str(identifier)).to_list()\n",
    "        attempts = 3\n",
    "        for project_id in bioproject_identifier:\n",
    "            # try querying NCBI three times\n",
    "            for attempt in range(1, attempts + 1):\n",
    "                try:\n",
    "                    srr_links = Entrez.elink(dbfrom=\"bioproject\",db=\"sra\", id=project_id)\n",
    "                    srr_link_results = Entrez.read(srr_links)\n",
    "                    srr_links.close()\n",
    "                    break # exit attempt loop if function succeeded\n",
    "                except Exception as e:\n",
    "                    print(\"\\t[-] ERROR during elink procedure with exception: {}\".format(e))\n",
    "                    if attempt == attempts:\n",
    "                        raise Exception(e)\n",
    "                    else:\n",
    "                        print(\"\\t[*] WARNING: trying elink procedure again after 1 second of sleep ...\")\n",
    "                        time.sleep(1)\n",
    "                        \n",
    "            link_list = []\n",
    "            for link_set in srr_link_results:\n",
    "                if \"LinkSetDb\" in link_set.keys():\n",
    "                    links = link_set[\"LinkSetDb\"]\n",
    "                    for sset in links:\n",
    "                        if \"DbTo\" in sset.keys():\n",
    "                            if sset[\"DbTo\"] == \"sra\":\n",
    "                                link_list_sets = sset[\"Link\"]\n",
    "        \n",
    "                                for lset in link_list_sets:\n",
    "                                    new_sra_id = str(lset[\"Id\"])\n",
    "                                    if new_sra_id not in link_list:\n",
    "                                        link_list.append(new_sra_id)\n",
    "            if len(link_list) == 0:\n",
    "                print(\"[*] WARNING: no hits for: {}\".format(project_id))\n",
    "            else:\n",
    "        \n",
    "                bioprojects_to_sra[project_id] = link_list\n",
    "            print(\"\\t[*] INFO: found {} entries for project: {}\".format(len(bioprojects_to_sra[project_id]), project_id))\n",
    "        \n",
    "        print(\"[+] DONE fetching information for srr\")\n",
    "        return bioprojects_to_sra\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR linking bioprojects to srr with exception: {}\".format(e))\n",
    "\n",
    "# pipeline step 6 - fetch detailed information regarding SRA entries\n",
    "def fetch_detailed_information_for_sra_data(bioproject_to_sra_dictionary:Dict,bioproject_to_sra_result_file_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    This function iterates over a dictionary with bioproject identifiers as keys and a list of SRA entries as values. It then fetches the esummary\n",
    "    files by Entrez for each SRA entry.\n",
    "    \n",
    "    :param\n",
    "        bioproject_to_sra_dictionary: Dict: input dictionary for iteration over bioproject identifier\n",
    "    :param\n",
    "        bioproject_to_sra_result_file_path: str: file path to output file\n",
    "\n",
    "    :returns\n",
    "        pd.DataFrame: result table\n",
    "    '''\n",
    "    try:\n",
    "        print(\"[*] Building result dataframe file\")\n",
    "        with open(bioproject_to_sra_result_file_path,\"w\") as outfile:\n",
    "            outfile.write(\"BioProjectId\\tTitle\\tPlatform\\tInstrument\\tRuns\\tSpots\\tBases\\tSubmitter\\tExperiment\\tStudy\\tOrganism\\tSample\\tLibrary\\tStrategy\\tSource\\tLayout\\tBioproject\\tBiosample\\n\")\n",
    "            for bioproject in bioproject_to_sra_dictionary.keys():\n",
    "                print(\"[*] Fetching detailed SRA information for bioproject: {}\".format(bioproject))\n",
    "                fetch = Entrez.esummary(db=\"sra\", id=\",\".join(bioproject_to_sra_dictionary[bioproject]),rettype=\"text\")\n",
    "                summaries = Entrez.read(fetch)\n",
    "                fetch.close()\n",
    "                #print(\"\\t[*] DONE fetching information\")\n",
    "                #print(\"\\t[*] Processing information ...\")\n",
    "                if len(summaries) != len(bioproject_to_sra_dictionary[bioproject]):\n",
    "                    print(\"\\t[!] Warning: length of summaries does not correspond to length of sra data: {} vs. {}\".format(len(summaries),len(bioproject_to_sra_dictionary[bioproject])))\n",
    "                for summary in summaries:\n",
    "                    xml_string = summary[\"ExpXml\"]\n",
    "                    xml_string = \"<root>\" + xml_string + \"</root>\"\n",
    "                    root = ET.fromstring(xml_string)# Extract information\n",
    "                    summ = root.find(\"Summary\")\n",
    "                    if summ is not None:\n",
    "                        title = summ.findtext(\"Title\", default=\"\")\n",
    "                        platform = summ.findtext(\"Platform\", default=\"\")\n",
    "                        platform_elem = summ.find(\"Platform\")\n",
    "                        instrument_model = platform_elem.attrib.get(\"instrument_model\") if platform_elem is not None else \"\"\n",
    "                        stats_elem = summ.find(\"Statistics\")\n",
    "                        if stats_elem is not None:\n",
    "                            total_runs = stats_elem.attrib.get(\"total_runs\", \"\")\n",
    "                            total_spots = stats_elem.attrib.get(\"total_spots\", \"\")\n",
    "                            total_bases = stats_elem.attrib.get(\"total_bases\", \"\")\n",
    "                        else:\n",
    "                            total_runs = total_spots = total_bases = \"\"\n",
    "                    else:\n",
    "                        title = platform = instrument_model = total_runs = total_spots = total_bases = \"\"\n",
    "        \n",
    "                    # Submitter, Experiment, Study, Organism, Sample\n",
    "                    submitter = root.find(\"Submitter\")\n",
    "                    submitter_acc = submitter.attrib.get(\"acc\", \"\") if submitter is not None else \"\"\n",
    "        \n",
    "                    experiment = root.find(\"Experiment\")\n",
    "                    experiment_acc = experiment.attrib.get(\"acc\", \"\") if experiment is not None else \"\"\n",
    "        \n",
    "                    study = root.find(\"Study\")\n",
    "                    study_acc = study.attrib.get(\"acc\", \"\") if study is not None else \"\"\n",
    "        \n",
    "                    organism = root.find(\"Organism\")\n",
    "                    organism_name = organism.attrib.get(\"ScientificName\", \"\") if organism is not None else \"\"\n",
    "        \n",
    "                    sample = root.find(\"Sample\")\n",
    "                    sample_acc = sample.attrib.get(\"acc\", \"\") if sample is not None else \"\"\n",
    "        \n",
    "                    # Library descriptor\n",
    "                    lib_descriptor = root.find(\"Library_descriptor\")\n",
    "                    if lib_descriptor is not None:\n",
    "                        library_name = lib_descriptor.findtext(\"LIBRARY_NAME\", default=\"\")\n",
    "                        library_strategy = lib_descriptor.findtext(\"LIBRARY_STRATEGY\", default=\"\")\n",
    "                        library_source = lib_descriptor.findtext(\"LIBRARY_SOURCE\", default=\"\")\n",
    "                        library_layout_elem = lib_descriptor.find(\"LIBRARY_LAYOUT\")\n",
    "                        if library_layout_elem is not None and len(library_layout_elem):\n",
    "                            library_layout = library_layout_elem[0].tag  # PAIRED or SINGLE\n",
    "                        else:\n",
    "                            library_layout = \"\"\n",
    "                    else:\n",
    "                        library_name = library_strategy = library_source = library_layout = \"\"\n",
    "        \n",
    "                    # BioProject and BioSample\n",
    "                    bioproject = root.findtext(\"Bioproject\", default=\"\")\n",
    "                    biosample = root.findtext(\"Biosample\", default=\"\")\n",
    "        \n",
    "                    line = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(bioproject,title,platform,instrument_model,\n",
    "                                                                                                        total_runs,total_spots,total_bases,\n",
    "                                                                                                        submitter_acc,experiment_acc,study_acc,\n",
    "                                                                                                        organism_name,sample_acc,library_name,\n",
    "                                                                                                        library_strategy,library_source,library_layout,\n",
    "                                                                                                        bioproject,biosample)\n",
    "                    outfile.write(line)\n",
    "                print(\"\\t[*] DONE\")\n",
    "            print(\"[+] DONE\")\n",
    "            return pd.read_table(bioproject_to_sra_result_file_path)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR fetching detailed information for SRA data with exception: {}\".format(e))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "17d0d6fb",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# helper functions\n",
    "def store_elink_mapping_data(sample_dictionary:Dict,output_filename:str, header:tuple, separator=\"\\t\")->pd.DataFrame:\n",
    "    '''\n",
    "    This function takes a dictionary containing Lists as values. It iterates over the keys and list-values and\n",
    "    outputs a CSV file containing the key - value information.\n",
    "\n",
    "    :param\n",
    "        sample_dictionary: Dict: dictionary containing a key and a list as value (e.g. bioproject identifier and corresponding sra links)\n",
    "    :param\n",
    "        output_filename: str: name and path of the output CSV file\n",
    "    :param\n",
    "        header: tuple(str, str): tuple of length two containing the key and value information as strings\n",
    "    :returns\n",
    "        pd.DataFrame\n",
    "    '''\n",
    "    try:\n",
    "        print(\"[*] Writing output file: {} to disc\".format(output_filename))\n",
    "        if len(header) != 2:\n",
    "            raise Exception(\"[-] ERROR as the header has not a size of two, choose something like: BioProjectID, SRA\")\n",
    "        else:\n",
    "            with open(output_filename, \"w\") as mapfile:\n",
    "                mapfile.write(header[0]+separator+header[1]+\"\\n\")\n",
    "                for key in sample_dictionary.keys():\n",
    "                    for val in sample_dictionary[key]:\n",
    "                        mapfile.write(key + separator + val + \"\\n\")\n",
    "        return pd.read_csv(output_filename, sep=separator)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR writing key - value file with exception: {}\".format(e))\n",
    "\n",
    "def read_csv_file(document_summary_csv:str, separator=\",\")->pd.DataFrame:\n",
    "    '''\n",
    "    Function for reading a csv table with pandas.\n",
    "    :param\n",
    "        document_summary_csv: str: Path to bioproject result file.\n",
    "        separator: str: Specified separator for the CSV file.\n",
    "    :return:\n",
    "        pd.DataFrame\n",
    "    '''\n",
    "    try:\n",
    "        return pd.read_csv(document_summary_csv, sep=separator)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR opening pandas table with exception: {}\".format(e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a9ecabc0",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Directory already exist.\n",
      "[*] Info: Found 13 BioProject's associated to Exaiptasia microbiome projects.\n",
      "[*] Info: Found detailed information for: 13 associated BioProject's\n",
      "[*] INFO: done writing result file with 13 entries.\n"
     ]
    }
   ],
   "source": [
    "query = '(\"Exaiptasia diaphana\"[Organism] AND microbiome)'\n",
    "bioproject_detailed_info_path = \"../results/exaiptasia_microbiome_studies.csv\"\n",
    "\n",
    "setup_entrez(\"lukas.becker@hhu.de\")\n",
    "setup_result_directory(\"../results/\")\n",
    "# step 1 get bioproject identifier\n",
    "bioproject_target_ids = fetch_bioprojects(query=query)\n",
    "# step 2 get detailed information\n",
    "bioproject_document_summary = fetch_detailed_bioproject_infos(bioproject_target_ids)\n",
    "# step 3 write result file\n",
    "write_bioproject_result_file(bioproject_detailed_info_path,bioproject_document_summary)\n",
    "bioproject_dataframe = read_csv_file(bioproject_detailed_info_path, separator=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e9b778f1",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ProjectId</th>\n",
       "      <th>ProjectAcc</th>\n",
       "      <th>ProjectDate</th>\n",
       "      <th>ProjectTitle</th>\n",
       "      <th>ProjectDescription</th>\n",
       "      <th>OrganismName</th>\n",
       "      <th>OrganismStrain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1364369</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>2025/11/16 00:00</td>\n",
       "      <td>Exaiptasia longitudinal microbiomes across tem...</td>\n",
       "      <td>Under global warming scenarios, the deteriorat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1336731</td>\n",
       "      <td>PRJNA1336731</td>\n",
       "      <td>2025/09/30 00:00</td>\n",
       "      <td>Symbiotic state modulates microbiome recovery ...</td>\n",
       "      <td>This study explored microbiome depletion and r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1089063</td>\n",
       "      <td>PRJNA1089063</td>\n",
       "      <td>2024/03/18 00:00</td>\n",
       "      <td>Exaiptasia diaphana strain:F003 | isolate:F003...</td>\n",
       "      <td>Coral reefs depend upon a functional symbiosis...</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>F003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>988282</td>\n",
       "      <td>PRJNA988282</td>\n",
       "      <td>2023/06/28 00:00</td>\n",
       "      <td>Microbiome of Exaiptasia diaphana and Berghia ...</td>\n",
       "      <td>The cladobranch Berghia stephanieae is able to...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>907389</td>\n",
       "      <td>PRJNA907389</td>\n",
       "      <td>2022/12/01 00:00</td>\n",
       "      <td>Bacterial microbiome variation across body par...</td>\n",
       "      <td>16s rRNA sequencing of bacterial communities i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ProjectId    ProjectAcc       ProjectDate  \\\n",
       "0    1364369   PRJEB101302  2025/11/16 00:00   \n",
       "1    1336731  PRJNA1336731  2025/09/30 00:00   \n",
       "2    1089063  PRJNA1089063  2024/03/18 00:00   \n",
       "3     988282   PRJNA988282  2023/06/28 00:00   \n",
       "4     907389   PRJNA907389  2022/12/01 00:00   \n",
       "\n",
       "                                        ProjectTitle  \\\n",
       "0  Exaiptasia longitudinal microbiomes across tem...   \n",
       "1  Symbiotic state modulates microbiome recovery ...   \n",
       "2  Exaiptasia diaphana strain:F003 | isolate:F003...   \n",
       "3  Microbiome of Exaiptasia diaphana and Berghia ...   \n",
       "4  Bacterial microbiome variation across body par...   \n",
       "\n",
       "                                  ProjectDescription         OrganismName  \\\n",
       "0  Under global warming scenarios, the deteriorat...                  NaN   \n",
       "1  This study explored microbiome depletion and r...                  NaN   \n",
       "2  Coral reefs depend upon a functional symbiosis...  Exaiptasia diaphana   \n",
       "3  The cladobranch Berghia stephanieae is able to...                  NaN   \n",
       "4  16s rRNA sequencing of bacterial communities i...                  NaN   \n",
       "\n",
       "  OrganismStrain  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2           F003  \n",
       "3            NaN  \n",
       "4            NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# result view\n",
    "bioproject_dataframe.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "97e5942c-796c-4323-8408-486ad82ecc0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Linking bioproject identifier to biosample database.\n",
      "\t[*] Found 590 entries for project: 1364369.\n",
      "\t[*] Found 88 entries for project: 1336731.\n",
      "\t[*] Found 2 entries for project: 1089063.\n",
      "\t[*] Found 5 entries for project: 988282.\n",
      "\t[*] Found 96 entries for project: 907389.\n",
      "\t[*] Found 14 entries for project: 650220.\n",
      "\t[*] Found 3 entries for project: 630329.\n",
      "\t[*] Found 6 entries for project: 592182.\n",
      "\t[*] Found 16 entries for project: 588472.\n",
      "\t[*] Found 4 entries for project: 576556.\n",
      "\t[*] Found 4 entries for project: 576020.\n",
      "\t[*] Found 58 entries for project: 524291.\n",
      "\t[*] Found 6 entries for project: 360672.\n",
      "[+] DONE fetching information for biosamples\n"
     ]
    }
   ],
   "source": [
    "# step 4 fetch biosample information - if you get a NCBI C++ error repeat executing this function\n",
    "bioprojects_to_biosamples_dictionary = get_biosample_from_bioprojects(bioproject_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "204245de-f6a6-46b2-90a4-c6b7b31ce9b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Linking bioproject identifier to srr database\n",
      "\t[*] INFO: found 590 entries for project: 1364369\n",
      "\t[*] INFO: found 88 entries for project: 1336731\n",
      "\t[*] INFO: found 2 entries for project: 1089063\n",
      "\t[*] INFO: found 26 entries for project: 988282\n",
      "\t[*] INFO: found 48 entries for project: 907389\n",
      "\t[*] INFO: found 14 entries for project: 650220\n",
      "\t[*] INFO: found 1071 entries for project: 630329\n",
      "\t[*] INFO: found 6 entries for project: 592182\n",
      "\t[*] INFO: found 16 entries for project: 588472\n",
      "\t[*] INFO: found 378 entries for project: 576556\n",
      "\t[*] INFO: found 1510 entries for project: 576020\n",
      "\t[*] INFO: found 58 entries for project: 524291\n",
      "\t[*] INFO: found 6 entries for project: 360672\n",
      "[+] DONE fetching information for srr\n"
     ]
    }
   ],
   "source": [
    "# step 5 fetch srr information\n",
    "bioprojects_to_sra_dictionary = get_sra_from_bioprojects(bioproject_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c89b3146-711d-498c-ac9b-468701362d52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Writing output file: ../results/bioproject_to_biosample.table to disk\n",
      "[*] Writing output file: ../results/bioproject_to_sra.table to disk\n"
     ]
    }
   ],
   "source": [
    "# step 6 - write results to disc\n",
    "bioproject_to_biosample_path = \"../results/bioproject_to_biosample.table\"\n",
    "bioproject_to_sra_path = \"../results/bioproject_to_sra.table\"\n",
    "bioproject_to_biosample_df = store_elink_mapping_data(bioprojects_to_biosamples_dictionary, \n",
    "                                                      output_filename=bioproject_to_biosample_path,\n",
    "                                                      header=(\"BioProjectID\",\"BioSampleID\"))\n",
    "bioproject_to_sra_df = store_elink_mapping_data(bioprojects_to_sra_dictionary, \n",
    "                                                output_filename=bioproject_to_sra_path,\n",
    "                                                header=(\"BioProjectID\",\"SraID\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bf84ddba-2332-457b-b354-33e9df0eb791",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Building result dataframe file\n",
      "[*] Fetching detailed SRA information for bioproject: 1364369\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 1336731\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 1089063\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 988282\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 907389\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 650220\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 630329\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 592182\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 588472\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 576556\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 576020\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 524291\n",
      "\t[*] DONE\n",
      "[*] Fetching detailed SRA information for bioproject: 360672\n",
      "\t[*] DONE\n",
      "[+] DONE\n"
     ]
    }
   ],
   "source": [
    "# step 7 - get detailed information of SRA entries\n",
    "bioproject_to_sra_detailed_file_path = \"../results/bioproject_to_sra_details.table\"\n",
    "bioproject_to_sra_detailed_table = fetch_detailed_information_for_sra_data(bioprojects_to_sra_dictionary, bioproject_to_sra_detailed_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3872a5ac-41c6-46c5-9213-262d6b83acab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProjectId</th>\n",
       "      <th>Title</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Runs</th>\n",
       "      <th>Spots</th>\n",
       "      <th>Bases</th>\n",
       "      <th>Submitter</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Study</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Library</th>\n",
       "      <th>Strategy</th>\n",
       "      <th>Source</th>\n",
       "      <th>Layout</th>\n",
       "      <th>Bioproject</th>\n",
       "      <th>Biosample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>Illumina MiSeq paired end sequencing</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>55499</td>\n",
       "      <td>27749500</td>\n",
       "      <td>ERA35072453</td>\n",
       "      <td>ERX15161653</td>\n",
       "      <td>ERP182725</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>ERS27119887</td>\n",
       "      <td>KMM_MiSeq_Aip</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>SAMEA120419201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>Illumina MiSeq paired end sequencing</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>53963</td>\n",
       "      <td>26981500</td>\n",
       "      <td>ERA35072453</td>\n",
       "      <td>ERX15161641</td>\n",
       "      <td>ERP182725</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>ERS27119874</td>\n",
       "      <td>KMM_MiSeq_Aip</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>SAMEA120419188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>Illumina MiSeq paired end sequencing</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>35839</td>\n",
       "      <td>17919500</td>\n",
       "      <td>ERA35072453</td>\n",
       "      <td>ERX15162068</td>\n",
       "      <td>ERP182725</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>ERS27119808</td>\n",
       "      <td>KMM_MiSeq_Aip</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>SAMEA120419122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>Illumina MiSeq paired end sequencing</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>62115</td>\n",
       "      <td>31057500</td>\n",
       "      <td>ERA35072453</td>\n",
       "      <td>ERX15161954</td>\n",
       "      <td>ERP182725</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>ERS27119681</td>\n",
       "      <td>KMM_MiSeq_Aip</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>SAMEA120418995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>Illumina MiSeq paired end sequencing</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>63187</td>\n",
       "      <td>31593500</td>\n",
       "      <td>ERA35072453</td>\n",
       "      <td>ERX15161693</td>\n",
       "      <td>ERP182725</td>\n",
       "      <td>Exaiptasia diaphana</td>\n",
       "      <td>ERS27119392</td>\n",
       "      <td>KMM_MiSeq_Aip</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PRJEB101302</td>\n",
       "      <td>SAMEA120418706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BioProjectId                                 Title  Platform  \\\n",
       "0  PRJEB101302  Illumina MiSeq paired end sequencing  ILLUMINA   \n",
       "1  PRJEB101302  Illumina MiSeq paired end sequencing  ILLUMINA   \n",
       "2  PRJEB101302  Illumina MiSeq paired end sequencing  ILLUMINA   \n",
       "3  PRJEB101302  Illumina MiSeq paired end sequencing  ILLUMINA   \n",
       "4  PRJEB101302  Illumina MiSeq paired end sequencing  ILLUMINA   \n",
       "\n",
       "       Instrument  Runs  Spots     Bases    Submitter   Experiment      Study  \\\n",
       "0  Illumina MiSeq     1  55499  27749500  ERA35072453  ERX15161653  ERP182725   \n",
       "1  Illumina MiSeq     1  53963  26981500  ERA35072453  ERX15161641  ERP182725   \n",
       "2  Illumina MiSeq     1  35839  17919500  ERA35072453  ERX15162068  ERP182725   \n",
       "3  Illumina MiSeq     1  62115  31057500  ERA35072453  ERX15161954  ERP182725   \n",
       "4  Illumina MiSeq     1  63187  31593500  ERA35072453  ERX15161693  ERP182725   \n",
       "\n",
       "              Organism       Sample        Library  Strategy       Source  \\\n",
       "0  Exaiptasia diaphana  ERS27119887  KMM_MiSeq_Aip  AMPLICON  METAGENOMIC   \n",
       "1  Exaiptasia diaphana  ERS27119874  KMM_MiSeq_Aip  AMPLICON  METAGENOMIC   \n",
       "2  Exaiptasia diaphana  ERS27119808  KMM_MiSeq_Aip  AMPLICON  METAGENOMIC   \n",
       "3  Exaiptasia diaphana  ERS27119681  KMM_MiSeq_Aip  AMPLICON  METAGENOMIC   \n",
       "4  Exaiptasia diaphana  ERS27119392  KMM_MiSeq_Aip  AMPLICON  METAGENOMIC   \n",
       "\n",
       "   Layout   Bioproject       Biosample  \n",
       "0  PAIRED  PRJEB101302  SAMEA120419201  \n",
       "1  PAIRED  PRJEB101302  SAMEA120419188  \n",
       "2  PAIRED  PRJEB101302  SAMEA120419122  \n",
       "3  PAIRED  PRJEB101302  SAMEA120418995  \n",
       "4  PAIRED  PRJEB101302  SAMEA120418706  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioproject_to_sra_detailed_table.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
