{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:10:55.246741400Z",
     "start_time": "2025-10-12T19:10:54.856520600Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "import xml.etree.ElementTree as ET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "# Entrez Setup\n",
    "Entrez.email = \"lukas.becker@hhu.de\""
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:10:55.287408100Z",
     "start_time": "2025-10-12T19:10:55.248741900Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [],
   "source": [
    "query = '(\"Exaiptasia diaphana\"[Organism] AND microbiome)'"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:35.754131700Z",
     "start_time": "2025-10-12T19:16:35.750127600Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [],
   "source": [
    "# 1 Query BioProject Database\n",
    "search_handle = Entrez.esearch(db=\"bioproject\", term=query, retmax=100)\n",
    "search_results = Entrez.read(search_handle)\n",
    "search_handle.close()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:37.263655900Z",
     "start_time": "2025-10-12T19:16:36.847007900Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Info: Found 12 BioProject's associated to Exaiptasia metagenomic projects.\n"
     ]
    }
   ],
   "source": [
    "print(\"[*] Info: Found {} BioProject's associated to Exaiptasia metagenomic projects.\".format(search_results[\"Count\"]))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:39.937148600Z",
     "start_time": "2025-10-12T19:16:39.935148100Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "# 2 Detailed searches based on fetched identifiers\n",
    "ids = search_results[\"IdList\"]\n",
    "handle = Entrez.efetch(db=\"bioproject\", id=\",\".join(ids), rettype=\"docsum\", retmode=\"xml\")\n",
    "projects = Entrez.read(handle)[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "handle.close()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:41.963930200Z",
     "start_time": "2025-10-12T19:16:41.399836200Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO: done writing result file with 12 entries.\n"
     ]
    }
   ],
   "source": [
    "# 3 Writing Results Into CSV Table\n",
    "bioprojects_resultfile = \"../data/bioproject_table.csv\"\n",
    "with open(bioprojects_resultfile,\"w\") as biofile:\n",
    "    header = \"ProjectId\\tProjectAcc\\tProjectDate\\tProjectTitle\\tProjectDescription\\tOrganismName\\tOrganismStrain\\n\"\n",
    "    biofile.write(header)\n",
    "    for project in projects:\n",
    "        project_id = project[\"Project_Id\"]\n",
    "        project_acc = project[\"Project_Acc\"]\n",
    "        project_date = project[\"Registration_Date\"]\n",
    "        project_title = project[\"Project_Title\"]\n",
    "        project_description = project[\"Project_Description\"]\n",
    "        project_organism = project[\"Organism_Name\"]\n",
    "        project_strain = project[\"Organism_Strain\"]\n",
    "        entry = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(project_id,\n",
    "                                                project_acc,\n",
    "                                                project_date,\n",
    "                                                project_title,\n",
    "                                                project_description,\n",
    "                                                project_organism,\n",
    "                                                project_strain)\n",
    "        biofile.write(entry)\n",
    "print(\"[*] INFO: done writing result file with {} entries.\".format(len(projects)))\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:48.276732900Z",
     "start_time": "2025-10-12T19:16:48.270732300Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "data": {
      "text/plain": "   ProjectId    ProjectAcc       ProjectDate  \\\n0    1336731  PRJNA1336731  2025/09/30 00:00   \n1    1089063  PRJNA1089063  2024/03/18 00:00   \n2     988282   PRJNA988282  2023/06/28 00:00   \n3     907389   PRJNA907389  2022/12/01 00:00   \n4     650220   PRJNA650220  2020/08/02 00:00   \n\n                                        ProjectTitle  \\\n0  Symbiotic state modulates microbiome recovery ...   \n1  Exaiptasia diaphana strain:F003 | isolate:F003...   \n2  Microbiome of Exaiptasia diaphana and Berghia ...   \n3  Bacterial microbiome variation across body par...   \n4  Microbiome characterization of Exaiptasia diap...   \n\n                                  ProjectDescription         OrganismName  \\\n0  This study explored microbiome depletion and r...                  NaN   \n1  Coral reefs depend upon a functional symbiosis...  Exaiptasia diaphana   \n2  The cladobranch Berghia stephanieae is able to...                  NaN   \n3  16s rRNA sequencing of bacterial communities i...                  NaN   \n4  Culture-dependent (isolation and 16S rRNA gene...  Exaiptasia diaphana   \n\n  OrganismStrain  \n0            NaN  \n1           F003  \n2            NaN  \n3            NaN  \n4            NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ProjectId</th>\n      <th>ProjectAcc</th>\n      <th>ProjectDate</th>\n      <th>ProjectTitle</th>\n      <th>ProjectDescription</th>\n      <th>OrganismName</th>\n      <th>OrganismStrain</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1336731</td>\n      <td>PRJNA1336731</td>\n      <td>2025/09/30 00:00</td>\n      <td>Symbiotic state modulates microbiome recovery ...</td>\n      <td>This study explored microbiome depletion and r...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1089063</td>\n      <td>PRJNA1089063</td>\n      <td>2024/03/18 00:00</td>\n      <td>Exaiptasia diaphana strain:F003 | isolate:F003...</td>\n      <td>Coral reefs depend upon a functional symbiosis...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>F003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>988282</td>\n      <td>PRJNA988282</td>\n      <td>2023/06/28 00:00</td>\n      <td>Microbiome of Exaiptasia diaphana and Berghia ...</td>\n      <td>The cladobranch Berghia stephanieae is able to...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>907389</td>\n      <td>PRJNA907389</td>\n      <td>2022/12/01 00:00</td>\n      <td>Bacterial microbiome variation across body par...</td>\n      <td>16s rRNA sequencing of bacterial communities i...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>650220</td>\n      <td>PRJNA650220</td>\n      <td>2020/08/02 00:00</td>\n      <td>Microbiome characterization of Exaiptasia diap...</td>\n      <td>Culture-dependent (isolation and 16S rRNA gene...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4 Check Dataframe\n",
    "df = pd.read_csv(bioprojects_resultfile,sep=\"\\t\")\n",
    "df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:16:52.909162700Z",
     "start_time": "2025-10-12T19:16:52.822428800Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO mapping bioproject accessions to biosample entries.\n",
      "\t[*] Working on BioProject: 1336731\n",
      "\t[*] Found 40 entries.\n",
      "\t[*] Working on BioProject: 1089063\n",
      "\t[*] Found 2 entries.\n",
      "\t[*] Working on BioProject: 988282\n",
      "\t[*] Found 5 entries.\n",
      "\t[*] Working on BioProject: 907389\n",
      "\t[*] Found 96 entries.\n",
      "\t[*] Working on BioProject: 650220\n",
      "\t[*] Found 14 entries.\n",
      "\t[*] Working on BioProject: 630329\n",
      "\t[*] Found 3 entries.\n",
      "\t[*] Working on BioProject: 592182\n",
      "\t[*] Found 6 entries.\n",
      "\t[*] Working on BioProject: 588472\n",
      "\t[*] Found 16 entries.\n",
      "\t[*] Working on BioProject: 576556\n",
      "\t[*] Found 4 entries.\n",
      "\t[*] Working on BioProject: 576020\n",
      "\t[*] Found 4 entries.\n",
      "\t[*] Working on BioProject: 524291\n",
      "\t[*] Found 58 entries.\n",
      "\t[*] Working on BioProject: 360672\n",
      "\t[*] Found 6 entries.\n",
      "[*] DONE\n"
     ]
    }
   ],
   "source": [
    "# 5 Get BioSample Data\n",
    "projects = df.ProjectId.apply(lambda x: str(x)).to_list()\n",
    "bioproject_to_biosample = {}\n",
    "print(\"[*] INFO mapping bioproject accessions to biosample entries.\")\n",
    "for proj in projects:\n",
    "    print(\"\\t[*] Working on BioProject: {}\".format(proj))\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\", db=\"biosample\", id=proj)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    biosample_ids = []\n",
    "    for link_set in link_results:\n",
    "        for link in link_set['LinkSetDb'][0]['Link']:\n",
    "            link_id = str(link[\"Id\"])\n",
    "            if link_id not in biosample_ids:\n",
    "                biosample_ids.append(link_id)\n",
    "    bioproject_to_biosample[proj] = biosample_ids\n",
    "    print(\"\\t[*] Found {} entries.\".format(len(biosample_ids)))\n",
    "print(\"[*] DONE\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:20:23.709767300Z",
     "start_time": "2025-10-12T19:20:17.428536500Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO: fetching SRA information for: 1336731.\n",
      "\t[*] INFO: found 40 entries.\n",
      "[*] INFO: fetching SRA information for: 1089063.\n",
      "\t[*] INFO: found 2 entries.\n",
      "[*] INFO: fetching SRA information for: 988282.\n",
      "\t[*] INFO: found 26 entries.\n",
      "[*] INFO: fetching SRA information for: 907389.\n",
      "\t[*] INFO: found 48 entries.\n",
      "[*] INFO: fetching SRA information for: 650220.\n",
      "\t[*] INFO: found 14 entries.\n",
      "[*] INFO: fetching SRA information for: 630329.\n",
      "\t[*] INFO: found 1071 entries.\n",
      "[*] INFO: fetching SRA information for: 592182.\n",
      "\t[*] INFO: found 6 entries.\n",
      "[*] INFO: fetching SRA information for: 588472.\n",
      "\t[*] INFO: found 16 entries.\n",
      "[*] INFO: fetching SRA information for: 576556.\n",
      "\t[*] INFO: found 378 entries.\n",
      "[*] INFO: fetching SRA information for: 576020.\n",
      "\t[*] INFO: found 1510 entries.\n",
      "[*] INFO: fetching SRA information for: 524291.\n",
      "\t[*] INFO: found 58 entries.\n",
      "[*] INFO: fetching SRA information for: 360672.\n",
      "\t[*] INFO: found 6 entries.\n",
      "[*] DONE parsing bioproject targets.\n"
     ]
    }
   ],
   "source": [
    "# 6 Get SRR data\n",
    "bioproject_to_srr_dict = {}\n",
    "for project_id in projects:\n",
    "    print(\"[*] INFO: fetching SRA information for: {}.\".format(project_id))\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\",db=\"sra\",id=project_id)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    link_list = []\n",
    "    for link_set in link_results:\n",
    "        if \"LinkSetDb\" in link_set.keys():\n",
    "            links = link_set[\"LinkSetDb\"]\n",
    "            for sset in links:\n",
    "                if \"DbTo\" in sset.keys():\n",
    "                    if sset[\"DbTo\"] == \"sra\":\n",
    "                        link_list_sets = sset[\"Link\"]\n",
    "\n",
    "                        for lset in link_list_sets:\n",
    "                            new_sra_id = str(lset[\"Id\"])\n",
    "                            if new_sra_id not in link_list:\n",
    "                                link_list.append(new_sra_id)\n",
    "    if len(link_list) == 0:\n",
    "        print(\"[*] WARNING: no hits for: {}\".format(project_id))\n",
    "    else:\n",
    "\n",
    "        bioproject_to_srr_dict[project_id] = link_list\n",
    "    print(\"\\t[*] INFO: found {} entries.\".format(len(bioproject_to_srr_dict[project_id])))\n",
    "print(\"[*] DONE parsing bioproject targets.\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:21:21.224477800Z",
     "start_time": "2025-10-12T19:21:13.788186200Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 7 Output Results Into Temporary File\n",
    "bioproject_srr_mapping_file = \"../data/bioproject_srr_table.csv\"\n",
    "with open(bioproject_srr_mapping_file,\"w\") as mapfile:\n",
    "    header = \"BioProject\\tSRR\\n\"\n",
    "    mapfile.write(header)\n",
    "    for bioproject in bioproject_to_srr_dict.keys():\n",
    "        for srr_link in bioproject_to_srr_dict[bioproject]:\n",
    "            mapfile.write(bioproject+\"\\t\"+srr_link+\"\\n\")\n",
    "bioproject_srr_df = pd.read_csv(bioproject_srr_mapping_file,sep=\"\\t\")\n",
    "bioproject_srr_df.head()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 8 Fetch Detailed Information\n",
    "print(\"[*] Building Result DataFrame File\")\n",
    "bioproject_srr_file = \"../data/bioproject_srr_details.csv\"\n",
    "with open(bioproject_srr_file,\"w\") as outfile:\n",
    "    outfile.write(\"BioProjectId\\tTitle\\tPlatform\\tInstrument\\tRuns\\tSpots\\tBases\\tSubmitter\\tExperiment\\tStudy\\tOrganism\\tSample\\tLibrary\\tStrategy\\tSource\\tLayout\\tBioproject\\tBiosample\\n\")\n",
    "    for bioproject in bioproject_to_srr_dict.keys():\n",
    "        print(\"[*] Fetching SRR information for bioproject: {}\".format(bioproject))\n",
    "        fetch = Entrez.esummary(db=\"sra\", id=\",\".join(bioproject_to_srr_dict[bioproject]),rettype=\"text\")\n",
    "        summaries = Entrez.read(fetch)\n",
    "        fetch.close()\n",
    "        print(\"\\t[*] DONE fetching information\")\n",
    "        print(\"\\t[*] Processing information ...\")\n",
    "        if len(summaries) != len(bioproject_to_srr_dict[bioproject]):\n",
    "            print(\"\\t[!] Warning: length of summaries does not correspond to length of srr data: {} vs. {}\".format(len(summaries),len(bioproject_to_srr_dict[bioproject])))\n",
    "        for summary in summaries:\n",
    "            xml_string = summary[\"ExpXml\"]\n",
    "            xml_string = \"<root>\" + xml_string + \"</root>\"\n",
    "            root = ET.fromstring(xml_string)# Extract information\n",
    "            summ = root.find(\"Summary\")\n",
    "            if summ is not None:\n",
    "                title = summ.findtext(\"Title\", default=\"\")\n",
    "                platform = summ.findtext(\"Platform\", default=\"\")\n",
    "                platform_elem = summ.find(\"Platform\")\n",
    "                instrument_model = platform_elem.attrib.get(\"instrument_model\") if platform_elem is not None else \"\"\n",
    "                stats_elem = summ.find(\"Statistics\")\n",
    "                if stats_elem is not None:\n",
    "                    total_runs = stats_elem.attrib.get(\"total_runs\", \"\")\n",
    "                    total_spots = stats_elem.attrib.get(\"total_spots\", \"\")\n",
    "                    total_bases = stats_elem.attrib.get(\"total_bases\", \"\")\n",
    "                else:\n",
    "                    total_runs = total_spots = total_bases = \"\"\n",
    "            else:\n",
    "                title = platform = instrument_model = total_runs = total_spots = total_bases = \"\"\n",
    "\n",
    "            # Submitter, Experiment, Study, Organism, Sample\n",
    "            submitter = root.find(\"Submitter\")\n",
    "            submitter_acc = submitter.attrib.get(\"acc\", \"\") if submitter is not None else \"\"\n",
    "\n",
    "            experiment = root.find(\"Experiment\")\n",
    "            experiment_acc = experiment.attrib.get(\"acc\", \"\") if experiment is not None else \"\"\n",
    "\n",
    "            study = root.find(\"Study\")\n",
    "            study_acc = study.attrib.get(\"acc\", \"\") if study is not None else \"\"\n",
    "\n",
    "            organism = root.find(\"Organism\")\n",
    "            organism_name = organism.attrib.get(\"ScientificName\", \"\") if organism is not None else \"\"\n",
    "\n",
    "            sample = root.find(\"Sample\")\n",
    "            sample_acc = sample.attrib.get(\"acc\", \"\") if sample is not None else \"\"\n",
    "\n",
    "            # Library descriptor\n",
    "            lib_descriptor = root.find(\"Library_descriptor\")\n",
    "            if lib_descriptor is not None:\n",
    "                library_name = lib_descriptor.findtext(\"LIBRARY_NAME\", default=\"\")\n",
    "                library_strategy = lib_descriptor.findtext(\"LIBRARY_STRATEGY\", default=\"\")\n",
    "                library_source = lib_descriptor.findtext(\"LIBRARY_SOURCE\", default=\"\")\n",
    "                library_layout_elem = lib_descriptor.find(\"LIBRARY_LAYOUT\")\n",
    "                if library_layout_elem is not None and len(library_layout_elem):\n",
    "                    library_layout = library_layout_elem[0].tag  # PAIRED or SINGLE\n",
    "                else:\n",
    "                    library_layout = \"\"\n",
    "            else:\n",
    "                library_name = library_strategy = library_source = library_layout = \"\"\n",
    "\n",
    "            # BioProject and BioSample\n",
    "            bioproject = root.findtext(\"Bioproject\", default=\"\")\n",
    "            biosample = root.findtext(\"Biosample\", default=\"\")\n",
    "\n",
    "            line = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(bioproject,title,platform,instrument_model,\n",
    "                                                                                                 total_runs,total_spots,total_bases,\n",
    "                                                                                                 submitter_acc,experiment_acc,study_acc,\n",
    "                                                                                                 organism_name,sample_acc,library_name,library_strategy,library_source,library_layout,bioproject,biosample)\n",
    "            outfile.write(line)\n",
    "        print(\"\\t[*] DONE\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "outputs": [
    {
     "data": {
      "text/plain": "   BioProjectId                                              Title  Platform  \\\n0  PRJNA1336731  RNA-seq: microbial depletion in Exaiptasia pal...  ILLUMINA   \n1  PRJNA1336731  RNA-seq: microbial depletion in Exaiptasia pal...  ILLUMINA   \n2  PRJNA1336731  RNA-seq: microbial depletion in Exaiptasia pal...  ILLUMINA   \n3  PRJNA1336731  RNA-seq: microbial depletion in Exaiptasia pal...  ILLUMINA   \n4  PRJNA1336731  RNA-seq: microbial depletion in Exaiptasia pal...  ILLUMINA   \n\n     Instrument  Runs    Spots      Bases   Submitter   Experiment      Study  \\\n0  NextSeq 1000     1  5928636  598792236  SRA2240826  SRX30732651  SRP630065   \n1  NextSeq 1000     1  4953153  500268453  SRA2240826  SRX30732650  SRP630065   \n2  NextSeq 1000     1  6247559  631003459  SRA2240826  SRX30732649  SRP630065   \n3  NextSeq 1000     1  4541768  458718568  SRA2240826  SRX30732648  SRP630065   \n4  NextSeq 1000     1  5721724  577894124  SRA2240826  SRX30732647  SRP630065   \n\n              Organism       Sample Library Strategy          Source  Layout  \\\n0  Exaiptasia diaphana  SRS26749035     CS2  RNA-Seq  TRANSCRIPTOMIC  SINGLE   \n1  Exaiptasia diaphana  SRS26749033     AS2  RNA-Seq  TRANSCRIPTOMIC  SINGLE   \n2  Exaiptasia diaphana  SRS26749032     AS3  RNA-Seq  TRANSCRIPTOMIC  SINGLE   \n3  Exaiptasia diaphana  SRS26749029     AS6  RNA-Seq  TRANSCRIPTOMIC  SINGLE   \n4  Exaiptasia diaphana  SRS26749031     AS7  RNA-Seq  TRANSCRIPTOMIC  SINGLE   \n\n     Bioproject     Biosample  \n0  PRJNA1336731  SAMN52082994  \n1  PRJNA1336731  SAMN52083003  \n2  PRJNA1336731  SAMN52083004  \n3  PRJNA1336731  SAMN52083005  \n4  PRJNA1336731  SAMN52083006  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>BioProjectId</th>\n      <th>Title</th>\n      <th>Platform</th>\n      <th>Instrument</th>\n      <th>Runs</th>\n      <th>Spots</th>\n      <th>Bases</th>\n      <th>Submitter</th>\n      <th>Experiment</th>\n      <th>Study</th>\n      <th>Organism</th>\n      <th>Sample</th>\n      <th>Library</th>\n      <th>Strategy</th>\n      <th>Source</th>\n      <th>Layout</th>\n      <th>Bioproject</th>\n      <th>Biosample</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>PRJNA1336731</td>\n      <td>RNA-seq: microbial depletion in Exaiptasia pal...</td>\n      <td>ILLUMINA</td>\n      <td>NextSeq 1000</td>\n      <td>1</td>\n      <td>5928636</td>\n      <td>598792236</td>\n      <td>SRA2240826</td>\n      <td>SRX30732651</td>\n      <td>SRP630065</td>\n      <td>Exaiptasia diaphana</td>\n      <td>SRS26749035</td>\n      <td>CS2</td>\n      <td>RNA-Seq</td>\n      <td>TRANSCRIPTOMIC</td>\n      <td>SINGLE</td>\n      <td>PRJNA1336731</td>\n      <td>SAMN52082994</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>PRJNA1336731</td>\n      <td>RNA-seq: microbial depletion in Exaiptasia pal...</td>\n      <td>ILLUMINA</td>\n      <td>NextSeq 1000</td>\n      <td>1</td>\n      <td>4953153</td>\n      <td>500268453</td>\n      <td>SRA2240826</td>\n      <td>SRX30732650</td>\n      <td>SRP630065</td>\n      <td>Exaiptasia diaphana</td>\n      <td>SRS26749033</td>\n      <td>AS2</td>\n      <td>RNA-Seq</td>\n      <td>TRANSCRIPTOMIC</td>\n      <td>SINGLE</td>\n      <td>PRJNA1336731</td>\n      <td>SAMN52083003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>PRJNA1336731</td>\n      <td>RNA-seq: microbial depletion in Exaiptasia pal...</td>\n      <td>ILLUMINA</td>\n      <td>NextSeq 1000</td>\n      <td>1</td>\n      <td>6247559</td>\n      <td>631003459</td>\n      <td>SRA2240826</td>\n      <td>SRX30732649</td>\n      <td>SRP630065</td>\n      <td>Exaiptasia diaphana</td>\n      <td>SRS26749032</td>\n      <td>AS3</td>\n      <td>RNA-Seq</td>\n      <td>TRANSCRIPTOMIC</td>\n      <td>SINGLE</td>\n      <td>PRJNA1336731</td>\n      <td>SAMN52083004</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>PRJNA1336731</td>\n      <td>RNA-seq: microbial depletion in Exaiptasia pal...</td>\n      <td>ILLUMINA</td>\n      <td>NextSeq 1000</td>\n      <td>1</td>\n      <td>4541768</td>\n      <td>458718568</td>\n      <td>SRA2240826</td>\n      <td>SRX30732648</td>\n      <td>SRP630065</td>\n      <td>Exaiptasia diaphana</td>\n      <td>SRS26749029</td>\n      <td>AS6</td>\n      <td>RNA-Seq</td>\n      <td>TRANSCRIPTOMIC</td>\n      <td>SINGLE</td>\n      <td>PRJNA1336731</td>\n      <td>SAMN52083005</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PRJNA1336731</td>\n      <td>RNA-seq: microbial depletion in Exaiptasia pal...</td>\n      <td>ILLUMINA</td>\n      <td>NextSeq 1000</td>\n      <td>1</td>\n      <td>5721724</td>\n      <td>577894124</td>\n      <td>SRA2240826</td>\n      <td>SRX30732647</td>\n      <td>SRP630065</td>\n      <td>Exaiptasia diaphana</td>\n      <td>SRS26749031</td>\n      <td>AS7</td>\n      <td>RNA-Seq</td>\n      <td>TRANSCRIPTOMIC</td>\n      <td>SINGLE</td>\n      <td>PRJNA1336731</td>\n      <td>SAMN52083006</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 9 Check Dataframe\n",
    "bioproject_srr_file = \"../data/bioproject_srr_details.csv\"\n",
    "srr_table = pd.read_csv(bioproject_srr_file,sep=\"\\t\")\n",
    "srr_table.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:22:00.572350500Z",
     "start_time": "2025-10-12T19:22:00.548681700Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO number os samples for: PRJNA1336731 is 40\n",
      "[*] INFO number os samples for: PRJNA1089063 is 2\n",
      "[*] INFO number os samples for: PRJNA988271 is 1\n",
      "[*] INFO number os samples for: PRJNA988282 is 25\n",
      "[*] INFO number os samples for: PRJNA907389 is 48\n",
      "[*] INFO number os samples for: PRJNA650220 is 14\n",
      "[*] INFO number os samples for: PRJNA630329 is 2142\n",
      "[*] INFO number os samples for: PRJNA592182 is 6\n",
      "[*] INFO number os samples for: PRJNA588472 is 16\n",
      "[*] INFO number os samples for: PRJNA576556 is 756\n",
      "[*] INFO number os samples for: PRJNA630325 is 61\n",
      "[*] INFO number os samples for: PRJNA524291 is 58\n",
      "[*] INFO number os samples for: PRJNA360672 is 6\n"
     ]
    }
   ],
   "source": [
    "for bioproject in srr_table.BioProjectId.unique():\n",
    "    print(\"[*] INFO number os samples for: {} is {}\".format(bioproject,len(srr_table[srr_table[\"BioProjectId\"] == bioproject])))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:22:38.856214700Z",
     "start_time": "2025-10-12T19:22:38.799921300Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "data": {
      "text/plain": "   ProjectId    ProjectAcc       ProjectDate  \\\n0    1336731  PRJNA1336731  2025/09/30 00:00   \n1    1089063  PRJNA1089063  2024/03/18 00:00   \n2     988282   PRJNA988282  2023/06/28 00:00   \n3     907389   PRJNA907389  2022/12/01 00:00   \n4     650220   PRJNA650220  2020/08/02 00:00   \n\n                                        ProjectTitle  \\\n0  Symbiotic state modulates microbiome recovery ...   \n1  Exaiptasia diaphana strain:F003 | isolate:F003...   \n2  Microbiome of Exaiptasia diaphana and Berghia ...   \n3  Bacterial microbiome variation across body par...   \n4  Microbiome characterization of Exaiptasia diap...   \n\n                                  ProjectDescription         OrganismName  \\\n0  This study explored microbiome depletion and r...                  NaN   \n1  Coral reefs depend upon a functional symbiosis...  Exaiptasia diaphana   \n2  The cladobranch Berghia stephanieae is able to...                  NaN   \n3  16s rRNA sequencing of bacterial communities i...                  NaN   \n4  Culture-dependent (isolation and 16S rRNA gene...  Exaiptasia diaphana   \n\n  OrganismStrain  \n0            NaN  \n1           F003  \n2            NaN  \n3            NaN  \n4            NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ProjectId</th>\n      <th>ProjectAcc</th>\n      <th>ProjectDate</th>\n      <th>ProjectTitle</th>\n      <th>ProjectDescription</th>\n      <th>OrganismName</th>\n      <th>OrganismStrain</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1336731</td>\n      <td>PRJNA1336731</td>\n      <td>2025/09/30 00:00</td>\n      <td>Symbiotic state modulates microbiome recovery ...</td>\n      <td>This study explored microbiome depletion and r...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1089063</td>\n      <td>PRJNA1089063</td>\n      <td>2024/03/18 00:00</td>\n      <td>Exaiptasia diaphana strain:F003 | isolate:F003...</td>\n      <td>Coral reefs depend upon a functional symbiosis...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>F003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>988282</td>\n      <td>PRJNA988282</td>\n      <td>2023/06/28 00:00</td>\n      <td>Microbiome of Exaiptasia diaphana and Berghia ...</td>\n      <td>The cladobranch Berghia stephanieae is able to...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>907389</td>\n      <td>PRJNA907389</td>\n      <td>2022/12/01 00:00</td>\n      <td>Bacterial microbiome variation across body par...</td>\n      <td>16s rRNA sequencing of bacterial communities i...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>650220</td>\n      <td>PRJNA650220</td>\n      <td>2020/08/02 00:00</td>\n      <td>Microbiome characterization of Exaiptasia diap...</td>\n      <td>Culture-dependent (isolation and 16S rRNA gene...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioprojects_resultfile = \"../data/bioproject_table.csv\"\n",
    "bioproject_info_df = pd.read_csv(bioprojects_resultfile,sep=\"\\t\")\n",
    "bioproject_info_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T19:14:25.792454100Z",
     "start_time": "2025-10-12T19:14:25.697032700Z"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
