{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:21.177433300Z",
     "start_time": "2025-10-12T13:28:20.838572500Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from xml.etree import ElementTree\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "# Entrez Setup\n",
    "Entrez.email = \"lukas.becker@hhu.de\""
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:21.179435Z",
     "start_time": "2025-10-12T13:28:21.177433300Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "query = '(\"Exaiptasia diaphana\"[Organism] AND microbiome)'"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:21.183268200Z",
     "start_time": "2025-10-12T13:28:21.180435100Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [],
   "source": [
    "# 1 Query BioProject Database\n",
    "search_handle = Entrez.esearch(db=\"bioproject\", term=query, retmax=100)\n",
    "search_results = Entrez.read(search_handle)\n",
    "search_handle.close()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:21.555089900Z",
     "start_time": "2025-10-12T13:28:21.184803500Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Info: Found 12 BioProject's associated to Exaiptasia metagenomic projects.\n"
     ]
    }
   ],
   "source": [
    "print(\"[*] Info: Found {} BioProject's associated to Exaiptasia metagenomic projects.\".format(search_results[\"Count\"]))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:21.559098900Z",
     "start_time": "2025-10-12T13:28:21.558094600Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [],
   "source": [
    "# 2 Detailed searches based on fetched identifiers\n",
    "ids = search_results[\"IdList\"]\n",
    "handle = Entrez.efetch(db=\"bioproject\", id=\",\".join(ids), rettype=\"docsum\", retmode=\"xml\")\n",
    "projects = Entrez.read(handle)[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "handle.close()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:22.119565600Z",
     "start_time": "2025-10-12T13:28:21.559098900Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO: done writing result file with 12 entries.\n"
     ]
    }
   ],
   "source": [
    "# 3 Writing Results Into CSV Table\n",
    "bioprojects_resultfile = \"../data/bioproject_table.csv\"\n",
    "with open(bioprojects_resultfile,\"w\") as biofile:\n",
    "    header = \"ProjectId\\tProjectAcc\\tProjectDate\\tProjectTitle\\tProjectDescription\\tOrganismName\\tOrganismStrain\\n\"\n",
    "    biofile.write(header)\n",
    "    for project in projects:\n",
    "        project_id = project[\"Project_Id\"]\n",
    "        project_acc = project[\"Project_Acc\"]\n",
    "        project_date = project[\"Registration_Date\"]\n",
    "        project_title = project[\"Project_Title\"]\n",
    "        project_description = project[\"Project_Description\"]\n",
    "        project_organism = project[\"Organism_Name\"]\n",
    "        project_strain = project[\"Organism_Strain\"]\n",
    "        entry = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(project_id,\n",
    "                                                project_acc,\n",
    "                                                project_date,\n",
    "                                                project_title,\n",
    "                                                project_description,\n",
    "                                                project_organism,\n",
    "                                                project_strain)\n",
    "        biofile.write(entry)\n",
    "print(\"[*] INFO: done writing result file with {} entries.\".format(len(projects)))\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:22.130731200Z",
     "start_time": "2025-10-12T13:28:22.124077800Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [
    {
     "data": {
      "text/plain": "   ProjectId    ProjectAcc       ProjectDate  \\\n0    1336731  PRJNA1336731  2025/09/30 00:00   \n1    1089063  PRJNA1089063  2024/03/18 00:00   \n2     988282   PRJNA988282  2023/06/28 00:00   \n3     907389   PRJNA907389  2022/12/01 00:00   \n4     650220   PRJNA650220  2020/08/02 00:00   \n\n                                        ProjectTitle  \\\n0  Symbiotic state modulates microbiome recovery ...   \n1  Exaiptasia diaphana strain:F003 | isolate:F003...   \n2  Microbiome of Exaiptasia diaphana and Berghia ...   \n3  Bacterial microbiome variation across body par...   \n4  Microbiome characterization of Exaiptasia diap...   \n\n                                  ProjectDescription         OrganismName  \\\n0  This study explored microbiome depletion and r...                  NaN   \n1  Coral reefs depend upon a functional symbiosis...  Exaiptasia diaphana   \n2  The cladobranch Berghia stephanieae is able to...                  NaN   \n3  16s rRNA sequencing of bacterial communities i...                  NaN   \n4  Culture-dependent (isolation and 16S rRNA gene...  Exaiptasia diaphana   \n\n  OrganismStrain  \n0            NaN  \n1           F003  \n2            NaN  \n3            NaN  \n4            NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ProjectId</th>\n      <th>ProjectAcc</th>\n      <th>ProjectDate</th>\n      <th>ProjectTitle</th>\n      <th>ProjectDescription</th>\n      <th>OrganismName</th>\n      <th>OrganismStrain</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1336731</td>\n      <td>PRJNA1336731</td>\n      <td>2025/09/30 00:00</td>\n      <td>Symbiotic state modulates microbiome recovery ...</td>\n      <td>This study explored microbiome depletion and r...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1089063</td>\n      <td>PRJNA1089063</td>\n      <td>2024/03/18 00:00</td>\n      <td>Exaiptasia diaphana strain:F003 | isolate:F003...</td>\n      <td>Coral reefs depend upon a functional symbiosis...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>F003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>988282</td>\n      <td>PRJNA988282</td>\n      <td>2023/06/28 00:00</td>\n      <td>Microbiome of Exaiptasia diaphana and Berghia ...</td>\n      <td>The cladobranch Berghia stephanieae is able to...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>907389</td>\n      <td>PRJNA907389</td>\n      <td>2022/12/01 00:00</td>\n      <td>Bacterial microbiome variation across body par...</td>\n      <td>16s rRNA sequencing of bacterial communities i...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>650220</td>\n      <td>PRJNA650220</td>\n      <td>2020/08/02 00:00</td>\n      <td>Microbiome characterization of Exaiptasia diap...</td>\n      <td>Culture-dependent (isolation and 16S rRNA gene...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4 Check Dataframe\n",
    "df = pd.read_csv(bioprojects_resultfile,sep=\"\\t\")\n",
    "df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:28:22.175937400Z",
     "start_time": "2025-10-12T13:28:22.130731200Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO mapping bioproject accessions to biosample entries.\n",
      "\t[*] Found 40 entries.\n",
      "\t[*] Found 2 entries.\n",
      "\t[*] Found 5 entries.\n",
      "\t[*] Found 96 entries.\n",
      "\t[*] Found 14 entries.\n",
      "\t[*] Found 3 entries.\n",
      "\t[*] Found 6 entries.\n",
      "\t[*] Found 16 entries.\n",
      "\t[*] Found 4 entries.\n",
      "\t[*] Found 4 entries.\n",
      "\t[*] Found 58 entries.\n",
      "\t[*] Found 6 entries.\n",
      "[*] DONE\n"
     ]
    }
   ],
   "source": [
    "# 5 Get BioSample Data\n",
    "projects = df.ProjectId.apply(lambda x: str(x)).to_list()\n",
    "bioproject_to_biosample = {}\n",
    "print(\"[*] INFO mapping bioproject accessions to biosample entries.\")\n",
    "for proj in projects:\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\", db=\"biosample\", id=proj)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    biosample_ids = []\n",
    "    for link_set in link_results:\n",
    "        for link in link_set['LinkSetDb'][0]['Link']:\n",
    "            link_id = str(link[\"Id\"])\n",
    "            if link_id not in biosample_ids:\n",
    "                biosample_ids.append(link_id)\n",
    "    bioproject_to_biosample[proj] = biosample_ids\n",
    "    print(\"\\t[*] Found {} entries.\".format(len(biosample_ids)))\n",
    "print(\"[*] DONE\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T13:41:17.450639600Z",
     "start_time": "2025-10-12T13:41:10.374071600Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] INFO: fetching SRA information for: 1336731.\n",
      "\t[*] INFO: found 40 entries.\n",
      "[*] INFO: fetching SRA information for: 1089063.\n",
      "\t[*] INFO: found 2 entries.\n",
      "[*] INFO: fetching SRA information for: 988282.\n",
      "\t[*] INFO: found 26 entries.\n",
      "[*] INFO: fetching SRA information for: 907389.\n",
      "\t[*] INFO: found 48 entries.\n",
      "[*] INFO: fetching SRA information for: 650220.\n",
      "\t[*] INFO: found 14 entries.\n",
      "[*] INFO: fetching SRA information for: 630329.\n",
      "\t[*] INFO: found 1071 entries.\n",
      "[*] INFO: fetching SRA information for: 592182.\n",
      "\t[*] INFO: found 6 entries.\n",
      "[*] INFO: fetching SRA information for: 588472.\n",
      "\t[*] INFO: found 16 entries.\n",
      "[*] INFO: fetching SRA information for: 576556.\n",
      "\t[*] INFO: found 378 entries.\n",
      "[*] INFO: fetching SRA information for: 576020.\n",
      "\t[*] INFO: found 1510 entries.\n",
      "[*] INFO: fetching SRA information for: 524291.\n",
      "\t[*] INFO: found 58 entries.\n",
      "[*] INFO: fetching SRA information for: 360672.\n",
      "\t[*] INFO: found 6 entries.\n",
      "[*] DONE parsing bioproject targets.\n"
     ]
    }
   ],
   "source": [
    "# 6 Get SRR data\n",
    "bioproject_to_srr_dict = {}\n",
    "for project_id in projects:\n",
    "    print(\"[*] INFO: fetching SRA information for: {}.\".format(project_id))\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\",db=\"sra\",id=project_id)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    link_list = []\n",
    "    for link_set in link_results:\n",
    "        if \"LinkSetDb\" in link_set.keys():\n",
    "            links = link_set[\"LinkSetDb\"]\n",
    "            for sset in links:\n",
    "                if \"DbTo\" in sset.keys():\n",
    "                    if sset[\"DbTo\"] == \"sra\":\n",
    "                        link_list_sets = sset[\"Link\"]\n",
    "\n",
    "                        for lset in link_list_sets:\n",
    "                            new_sra_id = str(lset[\"Id\"])\n",
    "                            if new_sra_id not in link_list:\n",
    "                                link_list.append(new_sra_id)\n",
    "    if len(link_list) == 0:\n",
    "        print(\"[*] WARNING: no hits for: {}\".format(project_id))\n",
    "    else:\n",
    "\n",
    "        bioproject_to_srr_dict[project_id] = link_list\n",
    "    print(\"\\t[*] INFO: found {} entries.\".format(len(bioproject_to_srr_dict[project_id])))\n",
    "print(\"[*] DONE parsing bioproject targets.\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T14:09:21.430340700Z",
     "start_time": "2025-10-12T14:09:13.300057500Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "outputs": [
    {
     "data": {
      "text/plain": "   BioProject       SRR\n0     1336731  41020117\n1     1336731  41020116\n2     1336731  41020115\n3     1336731  41020114\n4     1336731  41020113",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>BioProject</th>\n      <th>SRR</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1336731</td>\n      <td>41020117</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1336731</td>\n      <td>41020116</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1336731</td>\n      <td>41020115</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>1336731</td>\n      <td>41020114</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>1336731</td>\n      <td>41020113</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 7 Output Results Into Temporary File\n",
    "bioproject_srr_mapping_file = \"../data/bioproject_srr_table.csv\"\n",
    "with open(bioproject_srr_mapping_file,\"w\") as mapfile:\n",
    "    header = \"BioProject\\tSRR\\n\"\n",
    "    mapfile.write(header)\n",
    "    for bioproject in bioproject_to_srr_dict.keys():\n",
    "        for srr_link in bioproject_to_srr_dict[bioproject]:\n",
    "            mapfile.write(bioproject+\"\\t\"+srr_link+\"\\n\")\n",
    "bioproject_srr_df = pd.read_csv(bioproject_srr_mapping_file,sep=\"\\t\")\n",
    "bioproject_srr_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-12T14:20:50.422009400Z",
     "start_time": "2025-10-12T14:20:50.345553500Z"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
