{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "from typing import Dict, List\n",
    "from Bio import Entrez\n",
    "from pathlib import Path"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [],
   "source": [
    "# setup\n",
    "def setup_result_directory(result_dir:str)->Path:\n",
    "    '''\n",
    "    Creates a directory by using a specified string as argument.\n",
    "\n",
    "    :param\n",
    "        result_dir: directory path as string\n",
    "    :return:\n",
    "        result_dir_path: Path based on the specified result_dir string argument.\n",
    "    '''\n",
    "    try:\n",
    "        result_dir_path = Path(result_dir)\n",
    "        result_dir_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "        return result_dir_path\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR creating result directory with exception: {}\".format(e))\n",
    "\n",
    "def setup_entrez(email:str) -> None:\n",
    "    '''\n",
    "    Setup for the biopython Entrez.email field.\n",
    "\n",
    "    :param\n",
    "        email: String variable describing the user email.\n",
    "    :return:\n",
    "    '''\n",
    "    Entrez.email = email"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "outputs": [],
   "source": [
    "# pipeline step 1 - query NCBI based on a search string\n",
    "def fetch_bioprojects(query:str, max_returns=100)->List[int]:\n",
    "    '''\n",
    "    Function to perform an Entrez search query on the bioproject database.\n",
    "\n",
    "    :param\n",
    "        query: search string\n",
    "    :param\n",
    "        max_returns: maximal number of bioprojects to return\n",
    "    :return:\n",
    "        bioproject_search_results[\"IdList\"]: List[int] of bioproject identifier that can be used for fetching detailed information\n",
    "    '''\n",
    "    try:\n",
    "        bioproject_search_handle = Entrez.esearch(db=\"bioproject\", term=query, retmax=max_returns)\n",
    "        bioproject_search_results = Entrez.read(bioproject_search_handle)\n",
    "        bioproject_search_handle.close()\n",
    "        print(\"[*] Info: Found {} BioProject's associated to Exaiptasia microbiome projects.\".format(bioproject_search_results[\"Count\"]))\n",
    "        return bioproject_search_results[\"IdList\"]\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during fetching bioproject data with exception: {}\".format(e))\n",
    "\n",
    "# pipeline step 2 - get detailed information based on Entrez bioproject identifier\n",
    "def fetch_detailed_bioproject_infos(bioproject_identifier:List[int])->List[Dict]:\n",
    "    '''\n",
    "    Function for extracting detailed information of bioproject entries.\n",
    "\n",
    "    :param\n",
    "        bioproject_identifier: List[int]: list of bioproject identifier\n",
    "    :return:\n",
    "        bioproject_docsum: List[Dict]: detailed information of the requested bioprojects.\n",
    "    '''\n",
    "    try:\n",
    "        to_fetch_ids = \",\".join(bioproject_identifier)\n",
    "        bioproject_docsum_handle = Entrez.efetch(db=\"bioproject\", id=to_fetch_ids, rettype=\"docsum\", retmode=\"xml\")\n",
    "        bioproject_docsum = Entrez.read(bioproject_docsum_handle)[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "        bioproject_docsum_handle.close()\n",
    "        print(\"[*] Info: Found detailed information for: {} associated BioProject's\".format(len(bioproject_docsum)))\n",
    "        return bioproject_docsum\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during fetching bioproject detailed informarion: {}\".format(e))\n",
    "\n",
    "# pipeline step 3 - writing results into CSV file\n",
    "def write_bioproject_result_file(bioprojects_resultfile:str,project_document_summary:List[Dict])->None:\n",
    "    '''\n",
    "    Function for extracting document summary information fields. The information are written into a CSV file.\n",
    "\n",
    "    :param\n",
    "        bioprojects_resultfile: str: path to result file\n",
    "    :param\n",
    "        project_document_summary: List[Dict]: Entrez document summary\n",
    "    :return:\n",
    "    '''\n",
    "    try:\n",
    "        with open(bioprojects_resultfile,\"w\") as biofile:\n",
    "            header = \"ProjectId\\tProjectAcc\\tProjectDate\\tProjectTitle\\tProjectDescription\\tOrganismName\\tOrganismStrain\\n\"\n",
    "            biofile.write(header)\n",
    "            for project in project_document_summary:\n",
    "                project_id = project[\"Project_Id\"]\n",
    "                project_acc = project[\"Project_Acc\"]\n",
    "                project_date = project[\"Registration_Date\"]\n",
    "                project_title = project[\"Project_Title\"]\n",
    "                project_description = project[\"Project_Description\"]\n",
    "                project_organism = project[\"Organism_Name\"]\n",
    "                project_strain = project[\"Organism_Strain\"]\n",
    "                entry = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(project_id,\n",
    "                                                        project_acc,\n",
    "                                                        project_date,\n",
    "                                                        project_title,\n",
    "                                                        project_description,\n",
    "                                                        project_organism,\n",
    "                                                        project_strain)\n",
    "                biofile.write(entry)\n",
    "        print(\"[*] INFO: done writing result file with {} entries.\".format(len(project_document_summary)))\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR during BioProject result file writing with exception: {}\".format(e))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "outputs": [],
   "source": [
    "def read_bioproject_information(bioproject_document_summary_csv:str, separator=\",\")->pd.DataFrame:\n",
    "    '''\n",
    "    Function for reading bioproject csv table with pandas.\n",
    "    :param\n",
    "        bioproject_document_summary_csv: Path to bioproject result file.\n",
    "    :return:\n",
    "        pd.DataFrame\n",
    "    '''\n",
    "    try:\n",
    "        return pd.read_csv(bioproject_document_summary_csv, sep=separator)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"[-] ERROR opening pandas table with exception: {}\".format(e))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Info: Found 13 BioProject's associated to Exaiptasia metagenomic projects.\n",
      "[*] Info: Found detailed information for: 13 associated BioProject's\n",
      "[*] INFO: done writing result file with 13 entries.\n"
     ]
    }
   ],
   "source": [
    "query = '(\"Exaiptasia diaphana\"[Organism] AND microbiome)'\n",
    "bioproject_detailed_info_path = \"../results/exaiptasia_microbiome_studies.csv\"\n",
    "\n",
    "setup_entrez(\"lukas.becker@hhu.de\")\n",
    "# step 1 get bioproject identifier\n",
    "bioproject_target_ids = fetch_bioprojects(query=query)\n",
    "# step 2 get detailed information\n",
    "bioproject_document_summary = fetch_detailed_bioproject_infos(bioproject_target_ids)\n",
    "# step 3 write result file\n",
    "write_bioproject_result_file(bioproject_detailed_info_path,bioproject_document_summary)\n",
    "bioproject_dataframe = read_bioproject_information(bioproject_detailed_info_path, separator=\"\\t\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "outputs": [
    {
     "data": {
      "text/plain": "    ProjectId    ProjectAcc       ProjectDate  \\\n0     1364369   PRJEB101302  2025/11/16 00:00   \n1     1336731  PRJNA1336731  2025/09/30 00:00   \n2     1089063  PRJNA1089063  2024/03/18 00:00   \n3      988282   PRJNA988282  2023/06/28 00:00   \n4      907389   PRJNA907389  2022/12/01 00:00   \n5      650220   PRJNA650220  2020/08/02 00:00   \n6      630329   PRJNA630329  2020/05/04 00:00   \n7      592182   PRJNA592182  2019/11/27 00:00   \n8      588472   PRJNA588472  2019/11/08 00:00   \n9      576556   PRJNA576556  2019/10/09 00:00   \n10     576020   PRJNA576020  2019/10/05 00:00   \n11     524291   PRJNA524291  2019/02/26 00:00   \n12     360672   PRJNA360672  2017/01/10 00:00   \n\n                                         ProjectTitle  \\\n0   Exaiptasia longitudinal microbiomes across tem...   \n1   Symbiotic state modulates microbiome recovery ...   \n2   Exaiptasia diaphana strain:F003 | isolate:F003...   \n3   Microbiome of Exaiptasia diaphana and Berghia ...   \n4   Bacterial microbiome variation across body par...   \n5   Microbiome characterization of Exaiptasia diap...   \n6   Assessment of a ROS-targeted bacterial probiot...   \n7   Comparison of gene expression in Exaiptasia pa...   \n8   Comparison of gene expression in symbiotic and...   \n9           Reduced microbiome of Exaiptasia diaphana   \n10  Exaiptasia diaphana from the Great Barrier Ree...   \n11  Salinity-conveyed thermotolerance in the coral...   \n12                Bacterial microbiome of H2 Aiptasia   \n\n                                   ProjectDescription         OrganismName  \\\n0   Under global warming scenarios, the deteriorat...                  NaN   \n1   This study explored microbiome depletion and r...                  NaN   \n2   Coral reefs depend upon a functional symbiosis...  Exaiptasia diaphana   \n3   The cladobranch Berghia stephanieae is able to...                  NaN   \n4   16s rRNA sequencing of bacterial communities i...                  NaN   \n5   Culture-dependent (isolation and 16S rRNA gene...  Exaiptasia diaphana   \n6   Probiotic inoculation is one of several interv...  Exaiptasia diaphana   \n7   Emergence of the symbiotic lifestyle fostered ...  Exaiptasia diaphana   \n8   Emergence of the symbiotic lifestyle fostered ...  Exaiptasia diaphana   \n9   16S rRNA metabarcoding of the V5-V6 region to ...                  NaN   \n10  The sea anemone, Exaiptasia diaphana, commonly...  Exaiptasia diaphana   \n11  Here we characterized bacterial community comp...  Exaiptasia diaphana   \n12  Bacterial community associated with the H2 str...  Exaiptasia diaphana   \n\n   OrganismStrain  \n0             NaN  \n1             NaN  \n2            F003  \n3             NaN  \n4             NaN  \n5             NaN  \n6         AIMS2-4  \n7             NaN  \n8             NaN  \n9             NaN  \n10        AIMS1-4  \n11            NaN  \n12            NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ProjectId</th>\n      <th>ProjectAcc</th>\n      <th>ProjectDate</th>\n      <th>ProjectTitle</th>\n      <th>ProjectDescription</th>\n      <th>OrganismName</th>\n      <th>OrganismStrain</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1364369</td>\n      <td>PRJEB101302</td>\n      <td>2025/11/16 00:00</td>\n      <td>Exaiptasia longitudinal microbiomes across tem...</td>\n      <td>Under global warming scenarios, the deteriorat...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1336731</td>\n      <td>PRJNA1336731</td>\n      <td>2025/09/30 00:00</td>\n      <td>Symbiotic state modulates microbiome recovery ...</td>\n      <td>This study explored microbiome depletion and r...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1089063</td>\n      <td>PRJNA1089063</td>\n      <td>2024/03/18 00:00</td>\n      <td>Exaiptasia diaphana strain:F003 | isolate:F003...</td>\n      <td>Coral reefs depend upon a functional symbiosis...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>F003</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>988282</td>\n      <td>PRJNA988282</td>\n      <td>2023/06/28 00:00</td>\n      <td>Microbiome of Exaiptasia diaphana and Berghia ...</td>\n      <td>The cladobranch Berghia stephanieae is able to...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>907389</td>\n      <td>PRJNA907389</td>\n      <td>2022/12/01 00:00</td>\n      <td>Bacterial microbiome variation across body par...</td>\n      <td>16s rRNA sequencing of bacterial communities i...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>650220</td>\n      <td>PRJNA650220</td>\n      <td>2020/08/02 00:00</td>\n      <td>Microbiome characterization of Exaiptasia diap...</td>\n      <td>Culture-dependent (isolation and 16S rRNA gene...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>630329</td>\n      <td>PRJNA630329</td>\n      <td>2020/05/04 00:00</td>\n      <td>Assessment of a ROS-targeted bacterial probiot...</td>\n      <td>Probiotic inoculation is one of several interv...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>AIMS2-4</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>592182</td>\n      <td>PRJNA592182</td>\n      <td>2019/11/27 00:00</td>\n      <td>Comparison of gene expression in Exaiptasia pa...</td>\n      <td>Emergence of the symbiotic lifestyle fostered ...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>588472</td>\n      <td>PRJNA588472</td>\n      <td>2019/11/08 00:00</td>\n      <td>Comparison of gene expression in symbiotic and...</td>\n      <td>Emergence of the symbiotic lifestyle fostered ...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>576556</td>\n      <td>PRJNA576556</td>\n      <td>2019/10/09 00:00</td>\n      <td>Reduced microbiome of Exaiptasia diaphana</td>\n      <td>16S rRNA metabarcoding of the V5-V6 region to ...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>576020</td>\n      <td>PRJNA576020</td>\n      <td>2019/10/05 00:00</td>\n      <td>Exaiptasia diaphana from the Great Barrier Ree...</td>\n      <td>The sea anemone, Exaiptasia diaphana, commonly...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>AIMS1-4</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>524291</td>\n      <td>PRJNA524291</td>\n      <td>2019/02/26 00:00</td>\n      <td>Salinity-conveyed thermotolerance in the coral...</td>\n      <td>Here we characterized bacterial community comp...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>360672</td>\n      <td>PRJNA360672</td>\n      <td>2017/01/10 00:00</td>\n      <td>Bacterial microbiome of H2 Aiptasia</td>\n      <td>Bacterial community associated with the H2 str...</td>\n      <td>Exaiptasia diaphana</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioproject_dataframe"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
