{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "import xml.etree.ElementTree as ET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Entrez Setup\n",
    "Entrez.email = \"lukas.becker@hhu.de\""
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "query = '(\"Exaiptasia diaphana\"[Organism] AND microbiome)'"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 1 Query BioProject Database\n",
    "search_handle = Entrez.esearch(db=\"bioproject\", term=query, retmax=100)\n",
    "search_results = Entrez.read(search_handle)\n",
    "search_handle.close()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "print(\"[*] Info: Found {} BioProject's associated to Exaiptasia metagenomic projects.\".format(search_results[\"Count\"]))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 2 Detailed searches based on fetched identifiers\n",
    "ids = search_results[\"IdList\"]\n",
    "handle = Entrez.efetch(db=\"bioproject\", id=\",\".join(ids), rettype=\"docsum\", retmode=\"xml\")\n",
    "projects = Entrez.read(handle)[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "handle.close()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 3 Writing Results Into CSV Table\n",
    "bioprojects_resultfile = \"../data/bioproject_table.csv\"\n",
    "with open(bioprojects_resultfile,\"w\") as biofile:\n",
    "    header = \"ProjectId\\tProjectAcc\\tProjectDate\\tProjectTitle\\tProjectDescription\\tOrganismName\\tOrganismStrain\\n\"\n",
    "    biofile.write(header)\n",
    "    for project in projects:\n",
    "        project_id = project[\"Project_Id\"]\n",
    "        project_acc = project[\"Project_Acc\"]\n",
    "        project_date = project[\"Registration_Date\"]\n",
    "        project_title = project[\"Project_Title\"]\n",
    "        project_description = project[\"Project_Description\"]\n",
    "        project_organism = project[\"Organism_Name\"]\n",
    "        project_strain = project[\"Organism_Strain\"]\n",
    "        entry = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(project_id,\n",
    "                                                project_acc,\n",
    "                                                project_date,\n",
    "                                                project_title,\n",
    "                                                project_description,\n",
    "                                                project_organism,\n",
    "                                                project_strain)\n",
    "        biofile.write(entry)\n",
    "print(\"[*] INFO: done writing result file with {} entries.\".format(len(projects)))\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 4 Check Dataframe\n",
    "df = pd.read_csv(bioprojects_resultfile,sep=\"\\t\")\n",
    "df.head()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 5 Get BioSample Data\n",
    "projects = df.ProjectId.apply(lambda x: str(x)).to_list()\n",
    "bioproject_to_biosample = {}\n",
    "print(\"[*] INFO mapping bioproject accessions to biosample entries.\")\n",
    "for proj in projects:\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\", db=\"biosample\", id=proj)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    biosample_ids = []\n",
    "    for link_set in link_results:\n",
    "        for link in link_set['LinkSetDb'][0]['Link']:\n",
    "            link_id = str(link[\"Id\"])\n",
    "            if link_id not in biosample_ids:\n",
    "                biosample_ids.append(link_id)\n",
    "    bioproject_to_biosample[proj] = biosample_ids\n",
    "    print(\"\\t[*] Found {} entries.\".format(len(biosample_ids)))\n",
    "print(\"[*] DONE\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 6 Get SRR data\n",
    "bioproject_to_srr_dict = {}\n",
    "for project_id in projects:\n",
    "    print(\"[*] INFO: fetching SRA information for: {}.\".format(project_id))\n",
    "    link = Entrez.elink(dbfrom=\"bioproject\",db=\"sra\",id=project_id)\n",
    "    link_results = Entrez.read(link)\n",
    "    link.close()\n",
    "\n",
    "    link_list = []\n",
    "    for link_set in link_results:\n",
    "        if \"LinkSetDb\" in link_set.keys():\n",
    "            links = link_set[\"LinkSetDb\"]\n",
    "            for sset in links:\n",
    "                if \"DbTo\" in sset.keys():\n",
    "                    if sset[\"DbTo\"] == \"sra\":\n",
    "                        link_list_sets = sset[\"Link\"]\n",
    "\n",
    "                        for lset in link_list_sets:\n",
    "                            new_sra_id = str(lset[\"Id\"])\n",
    "                            if new_sra_id not in link_list:\n",
    "                                link_list.append(new_sra_id)\n",
    "    if len(link_list) == 0:\n",
    "        print(\"[*] WARNING: no hits for: {}\".format(project_id))\n",
    "    else:\n",
    "\n",
    "        bioproject_to_srr_dict[project_id] = link_list\n",
    "    print(\"\\t[*] INFO: found {} entries.\".format(len(bioproject_to_srr_dict[project_id])))\n",
    "print(\"[*] DONE parsing bioproject targets.\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 7 Output Results Into Temporary File\n",
    "bioproject_srr_mapping_file = \"../data/bioproject_srr_table.csv\"\n",
    "with open(bioproject_srr_mapping_file,\"w\") as mapfile:\n",
    "    header = \"BioProject\\tSRR\\n\"\n",
    "    mapfile.write(header)\n",
    "    for bioproject in bioproject_to_srr_dict.keys():\n",
    "        for srr_link in bioproject_to_srr_dict[bioproject]:\n",
    "            mapfile.write(bioproject+\"\\t\"+srr_link+\"\\n\")\n",
    "bioproject_srr_df = pd.read_csv(bioproject_srr_mapping_file,sep=\"\\t\")\n",
    "bioproject_srr_df.head()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 8 Fetch Detailed Information\n",
    "print(\"[*] Building Result DataFrame File\")\n",
    "bioproject_srr_file = \"../data/bioproject_srr_details.csv\"\n",
    "with open(bioproject_srr_file,\"w\") as outfile:\n",
    "    outfile.write(\"BioProjectId\\tTitle\\tPlatform\\tInstrument\\tRuns\\tSpots\\tBases\\tSubmitter\\tExperiment\\tStudy\\tOrganism\\tSample\\tLibrary\\tStrategy\\tSource\\tLayout\\tBioproject\\tBiosample\\n\")\n",
    "    for bioproject in bioproject_to_srr_dict.keys():\n",
    "        print(\"[*] Fetching SRR information for bioproject: {}\".format(bioproject))\n",
    "        fetch = Entrez.esummary(db=\"sra\", id=\",\".join(bioproject_to_srr_dict[bioproject]),rettype=\"text\")\n",
    "        summaries = Entrez.read(fetch)\n",
    "        fetch.close()\n",
    "        print(\"\\t[*] DONE fetching information\")\n",
    "        print(\"\\t[*] Processing information ...\")\n",
    "        if len(summaries) != len(bioproject_to_srr_dict[bioproject]):\n",
    "            print(\"\\t[!] Warning: length of summaries does not correspond to length of srr data: {} vs. {}\".format(len(summaries),len(bioproject_to_srr_dict[bioproject])))\n",
    "        for summary in summaries:\n",
    "            xml_string = summary[\"ExpXml\"]\n",
    "            xml_string = \"<root>\" + xml_string + \"</root>\"\n",
    "            root = ET.fromstring(xml_string)# Extract information\n",
    "            summ = root.find(\"Summary\")\n",
    "            if summ is not None:\n",
    "                title = summ.findtext(\"Title\", default=\"\")\n",
    "                platform = summ.findtext(\"Platform\", default=\"\")\n",
    "                platform_elem = summ.find(\"Platform\")\n",
    "                instrument_model = platform_elem.attrib.get(\"instrument_model\") if platform_elem is not None else \"\"\n",
    "                stats_elem = summ.find(\"Statistics\")\n",
    "                if stats_elem is not None:\n",
    "                    total_runs = stats_elem.attrib.get(\"total_runs\", \"\")\n",
    "                    total_spots = stats_elem.attrib.get(\"total_spots\", \"\")\n",
    "                    total_bases = stats_elem.attrib.get(\"total_bases\", \"\")\n",
    "                else:\n",
    "                    total_runs = total_spots = total_bases = \"\"\n",
    "            else:\n",
    "                title = platform = instrument_model = total_runs = total_spots = total_bases = \"\"\n",
    "\n",
    "            # Submitter, Experiment, Study, Organism, Sample\n",
    "            submitter = root.find(\"Submitter\")\n",
    "            submitter_acc = submitter.attrib.get(\"acc\", \"\") if submitter is not None else \"\"\n",
    "\n",
    "            experiment = root.find(\"Experiment\")\n",
    "            experiment_acc = experiment.attrib.get(\"acc\", \"\") if experiment is not None else \"\"\n",
    "\n",
    "            study = root.find(\"Study\")\n",
    "            study_acc = study.attrib.get(\"acc\", \"\") if study is not None else \"\"\n",
    "\n",
    "            organism = root.find(\"Organism\")\n",
    "            organism_name = organism.attrib.get(\"ScientificName\", \"\") if organism is not None else \"\"\n",
    "\n",
    "            sample = root.find(\"Sample\")\n",
    "            sample_acc = sample.attrib.get(\"acc\", \"\") if sample is not None else \"\"\n",
    "\n",
    "            # Library descriptor\n",
    "            lib_descriptor = root.find(\"Library_descriptor\")\n",
    "            if lib_descriptor is not None:\n",
    "                library_name = lib_descriptor.findtext(\"LIBRARY_NAME\", default=\"\")\n",
    "                library_strategy = lib_descriptor.findtext(\"LIBRARY_STRATEGY\", default=\"\")\n",
    "                library_source = lib_descriptor.findtext(\"LIBRARY_SOURCE\", default=\"\")\n",
    "                library_layout_elem = lib_descriptor.find(\"LIBRARY_LAYOUT\")\n",
    "                if library_layout_elem is not None and len(library_layout_elem):\n",
    "                    library_layout = library_layout_elem[0].tag  # PAIRED or SINGLE\n",
    "                else:\n",
    "                    library_layout = \"\"\n",
    "            else:\n",
    "                library_name = library_strategy = library_source = library_layout = \"\"\n",
    "\n",
    "            # BioProject and BioSample\n",
    "            bioproject = root.findtext(\"Bioproject\", default=\"\")\n",
    "            biosample = root.findtext(\"Biosample\", default=\"\")\n",
    "\n",
    "            line = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(bioproject,title,platform,instrument_model,\n",
    "                                                                                                 total_runs,total_spots,total_bases,\n",
    "                                                                                                 submitter_acc,experiment_acc,study_acc,\n",
    "                                                                                                 organism_name,sample_acc,library_name,library_strategy,library_source,library_layout,bioproject,biosample)\n",
    "            outfile.write(line)\n",
    "        print(\"\\t[*] DONE\")\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# 9 Check Dataframe\n",
    "srr_table = pd.read_csv(bioproject_srr_file,sep=\"\\t\")\n",
    "srr_table.head()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "for bioproject in srr_table.BioProjectId.unique():\n",
    "    print(\"[*] INFO number os samples for: {} is {}\".format(bioproject,len(srr_table[srr_table[\"BioProjectId\"] == bioproject])))"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
